packages/@aws-cdk/aws-pipes-sources-alpha/lib/streamSource.ts

```diff
@@ -0,0 +1,143 @@
+   *
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipesourcekinesisstreamparameters.html#cfn-pipes-pipe-pipesourcekinesisstreamparameters-maximumretryattempts
+   * @default -1 - EventBridge will retry failed records until the record expires in the event source
+   */
+  readonly maximumRetryAttempts?: number;
+
+  /**
+   * Define how to handle item process failures. {@link OnPartialBatchItemFailure.AUTOMATIC_BISECT} halves each batch and will retry each half until all the records are processed or there is one failed message left in the batch.
+   *
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipesourcekinesisstreamparameters.html#cfn-pipes-pipe-pipesourcekinesisstreamparameters-onpartialbatchitemfailure
+   * @default off - EventBridge will retry the entire batch
+   */
+  readonly onPartialBatchItemFailure?: OnPartialBatchItemFailure;
+
+  /**
+   * The number of batches to process concurrently from each shard.
+   *
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-pipes-pipe-pipesourcekinesisstreamparameters.html#cfn-pipes-pipe-pipesourcekinesisstreamparameters-parallelizationfactor
+   * @default 1
+   */
+  readonly parallelizationFactor?: number;
+}
+
+/**
+ * Streaming sources.
+ */
+export abstract class StreamSource extends SourceWithDeadLetterTarget {
+  /**
+   * The ARN of the source resource.
+   */
+  readonly sourceArn: string;
+  /**
+   * Base parameters for streaming sources.
+   */
+  readonly sourceParameters: StreamSourceParameters;
+  /**
+   * The maximum length of a time to wait for events in seconds.
+   */
+  readonly maximumBatchingWindowInSeconds;
+  /**
+   * The maximum record age in seconds.
+   */
+  readonly maximumRecordAgeInSeconds;
+
+  constructor(sourceArn: string, sourceParameters: StreamSourceParameters) {
+    super(sourceArn, sourceParameters.deadLetterTarget);
+    this.sourceArn = sourceArn;
+    this.sourceParameters = sourceParameters;
+    this.maximumBatchingWindowInSeconds = this.sourceParameters.maximumBatchingWindow?.toSeconds();
+    this.maximumRecordAgeInSeconds = this.sourceParameters.maximumRecordAge?.toSeconds();
+
+    validateBatchSize(this.sourceParameters.batchSize);
+    validateMaximumBatchingWindow(this.maximumBatchingWindowInSeconds);
+    validateMaximumRecordAge(this.maximumRecordAgeInSeconds);
+    validateMaxiumRetryAttemps(this.sourceParameters.maximumRetryAttempts);
+    validateParallelizationFactor(this.sourceParameters.parallelizationFactor);
+  }
+}
+
+function validateBatchSize(batchSize?: number) {
+  if (batchSize !== undefined && !Token.isUnresolved(batchSize)) {
+    if (batchSize < 1 || batchSize > 10000) {
+      throw new Error(`Batch size must be between 1 and 10000, received ${batchSize}`);
+    }
+  }
+}
+
+function validateMaximumBatchingWindow(window?: number) {
+  if (window !== undefined) {
```

Comment:
since `maximumBatchingWindow` is a `Duration`, is there a good way to check for a `Token` here? Or can this check be skipped?

Reply:
I think we can skip it given the `window` is from `Duration.toSeconds`